model_compression_toolkit/common/mixed_precision/search_methods/linear_programming.py

```diff
@@ -176,27 +146,43 @@ def _formalize_problem(layer_to_indicator_vars_mapping: Dict[int, Dict[int, LpVa
-            if not np.isinf(target_kpi.activation_memory):
-                activation_by_indicators = [indicator * layer_to_kpi_mapping[layer][nbits].activation_memory
-                                            for nbits, indicator in layer_to_indicator_vars_mapping[layer].items()]
-                total_activation_consumption.extend(activation_by_indicators)
-
-        # Total model memory size is bounded to the given KPI.
-        # Since total_weights_consumption and total_activation_consumption is the contribution
-        # to the minimal possible KPI (for weights and activation respectively),
-        # we bound the problem by the difference of the target KPI to the minimal KPI.
+            for _, indicator in layer_to_indicator_vars_mapping[layer].items():
+                indicators.append(indicator)
+
+        indicators_arr = np.array(indicators)
+        indicators_matrix = np.diag(indicators_arr)
+
         if not np.isinf(target_kpi.weights_memory):
-            lp_problem += lpSum(total_weights_consumption) <= target_kpi.weights_memory - minimal_kpi.weights_memory
+            weights_kpi_metrix = search_manager.compute_kpi_metrix(KPITarget.WEIGHTS)
+            indicators_weights_kpi_matrix = np.matmul(weights_kpi_metrix, indicators_matrix)
+
+            weights_kpi_sum_vector = [
+                sum([indicators_weights_kpi_matrix[i][j] for j in range(indicators_weights_kpi_matrix.shape[1])]) +
+                search_manager.min_kpi[KPITarget.WEIGHTS][i] for i in range(indicators_weights_kpi_matrix.shape[0])]
+
+            # search_manager.compute_kpi_functions contains a pair of kpi_metric and kpi_aggregation for each kpi target
+            aggr_weights_kpi = search_manager.compute_kpi_functions[KPITarget.WEIGHTS][1](weights_kpi_sum_vector)
+
+            for v in aggr_weights_kpi:
+                lp_problem += v <= target_kpi.weights_memory
+
         if not np.isinf(target_kpi.activation_memory):
-            lp_problem += lpSum(
-                total_activation_consumption) <= target_kpi.activation_memory - minimal_kpi.activation_memory
+            activation_kpi_metrix = search_manager.compute_kpi_metrix(KPITarget.ACTIVATION)
+            indicators_activation_kpi_matrix = np.matmul(activation_kpi_metrix, indicators_matrix)
+
+            activation_kpi_sum_vector = [
+                sum([indicators_activation_kpi_matrix[i][j] for j in range(indicators_activation_kpi_matrix.shape[1])])
+                + search_manager.min_kpi[KPITarget.ACTIVATION][i] for i in range(indicators_activation_kpi_matrix.shape[0])]
+
+            # search_manager.compute_kpi_functions contains a pair of kpi_metric and kpi_aggregation for each kpi target
+            aggr_activation_kpi = \
+                search_manager.compute_kpi_functions[KPITarget.ACTIVATION][1](activation_kpi_sum_vector)
+
+            for v in aggr_activation_kpi:
+                lp_problem += v <= target_kpi.activation_memory
+
     else:
```

Comment:
The 2 blocks are similar. If we can attach a "user constraint attribute" to a KPITarget, we can probably shorten this.